### ISSUE TITLE
Improve kubelet metrics APIs

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#12483]
Currently, kubelet exposes container stats provided by cAdvisor via `/stats/` REST endpoint. This API provides per container stats only. It does not provide stats for the pod network container and does not expose system and daemon containers on the node. This API is also not versioned. 

This API needs to be replaced with versioned Pod and Node stats APIs, that exposes all the node and pod metrics including network usage.
Also, supporting downsampling at kubelet will help reduce the network load.

**Comment by vishh** [comment]
cc @mvdan who had started looking into this issue.

**Comment by vishh** [comment]
Once this issue is resolved, cAdvisor REST endpoint can be blocked by default, to prevent access to potentially sensitive user data.

**Comment by timothysc** [comment]
cc @derekwaynecarr

**Comment by jimmidyson** [comment]
Can we also update the Prometheus exposed metrics to be properly labelled with kube metadata?

**Comment by dchen1107** [comment]
@vishh I am assigning this to you since you are supervising intern on this.

**Comment by a-robinson** [comment]
Is this still on track to make 1.1?

**Comment by mvdan** [comment]
Note that we are re-using the metrics provided by the model in Heapster instead of feeding from improved kubelet APIs. So we are not working on this currently.

Heapster itself does feed from the current kubelet APIs, so improving the APIs and making Heapster use the new ones is just a future step to improve the efficiency of the stats gathering process.

**Comment by dchen1107** [comment]
Ok, I retarget this for 1.2.

**Comment by vishh** [comment]
Targeting this for v1.2 makes sense.

**Comment by vishh** [comment]
@dchen1107: I'm adding this issue to the v1.1 milestone based on the offline discussion.

**Comment by yujuhong** [comment]
> Also, supporting downsampling at kubelet will help reduce the network load.

Why can't we do this in cadvisor? I'd rather make it tunable in cadvisor than doing this in kubelet itself.

Does this issue also include the prometheus metrics/ endpoint or is it strictly restricted to stats/?

**Comment by vishh** [comment]
> Why can't we do this in cadvisor? I'd rather make it tunable in cadvisor than doing this in kubelet itself.

+1. This is a nice to have. Not required in the first iteration since just specifying a time range should work in most cases.

> Does this issue also include the prometheus metrics/ endpoint or is it strictly restricted to stats/?

I vote for moving prometheus specific work into a separate issue for now.

**Comment by jimmidyson** [comment]
I've already got stuff working with Prometheus metrics & Kubernetes labels since google/cadvisor#906 was merged - just need to get a bump on next cadvisor release & add in the custom mapping function. Works nicely.

**Comment by dchen1107** [comment]
I like what @yujuhong proposed above to do this in cadvisor.

**Comment by vishh** [comment]
The primary motivation for this new API as as follows:

1. The number of queries against the `/stats` endpoint is proportional to the number of containers. This adds a considerable performance and resource usage overhead to the kubelet. Another side effect is users paying more for system overhead in the case of deployments like GKE.
2. Heapster scalability.

Since the primary user of the new API is heapster, we can start by just aggregating [cAdvisor spec](https://github.com/google/cadvisor/blob/master/info/v2/container.go#L55) and [stats objects](https://github.com/google/cadvisor/blob/master/info/v2/container.go#L88) into pods and system containers.

I'm thinking the new endpoint can be `/api/v1/rawMetrics?start=time.Time{}&end=time.Time{}`. We can re-use cAdvisor API objects (links above) for serving this endpoint. 

```go
type ContainerMetrics struct {
        TypeMeta
        ObjectMeta // should contain container name
        Spec *cAdvisorApiv2.ContainerSpec
        Stats []*cAdvisorApiv2.ContainerStats
}
type PodMetrics struct {
        TypeMeta
        ObjectMeta // should contain pod name
        Containers []ContainerMetrics
}

type RawMetrics struct {
        TypeMeta
        ObjectMeta // should contain node name
        Machine ContainerMetrics // Overall machine usage
        SystemContainers []ContainerMetrics // System containers like /kubelet, /docker-daemon, etc.
        Pods []PodMetrics
}
```

We might want to add support for querying rawMetrics for a specific pod or a node, but that might not be necessary to start with.

We will eventually need to expose a [`derivedMetrics`](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/compute-resource-metrics-api.md) endpoint as well to move resource estimation to the node.  This is not necessary for now.

**Comment by dchen1107** [comment]
Let me clarify a little bit on above comment:

* We talked about introduce a plugin to cAdvisor so that it can understand the pods and containers relateship. But that is more implementation issue to make Kubelet more modularized. 
* No matter where to implement the logic, we need to clean up kubelet /stats/ endpoint anyway. I am not intended to open cAdvsor port to provide a direct access to cAdvsor API.

A little bit background on why we promote this back to 1.1 milestone:
* Issue https://github.com/kubernetes/kubernetes/issues/15315 was filed to enable HorizontalPodAutoscaler as beta feature, which was expected as experimental one, and disabled by default.
* To support this as beta feature, we have to make sure the path of pulling metrics is reliable. Today's design and implementation causes reliable, scalability issue, and potentially increase kubelet cpu usage.

Had more offline discussion with @vishh. For 1.1 release, we can simplify the work by providing stats for pod or top-level resource-only containers managed by kubelet through a separate endpoint.

**Comment by bgrant0607** [comment]
Punting to 1.2.

**Comment by davidopp** [comment]
> Punting to 1.2.

My understanding was that "we can simplify the work by providing stats for pod or top-level resource-only containers managed by kubelet through a separate endpoint" is a blocker for horizontal pod autoscaling. (But everything else in the issue can be punted.)

**Comment by timstclair** [comment]
FWIW, I'm close to having the PR for this ready (progress in https://github.com/kubernetes/kubernetes/pull/15691) - modulo review latency & flaky tests.

**Comment by davidopp** [comment]
@fgrzadkowski @dchen1107

**Comment by timstclair** [comment]
Status update:

- [ ]  https://github.com/kubernetes/kubernetes/pull/19932 - `/stats/summary` implementation **[LGTM]**
- [ ]  https://github.com/kubernetes/kubernetes/pull/19741 - disk usage stats implementation **[In review]**
- [ ]  https://github.com/kubernetes/heapster/pull/932 - update heapster Godeps with summary API **[In review]**
- [ ]  https://github.com/kubernetes/heapster/pull/927 - read from summary API in heapster **[WIP]**

**Comment by dchen1107** [comment]
Great! The only thing left is cut another heapster release which uses the new metrics API. :-)
